<a href="https://colab.research.google.com/github/payasvi/dl/blob/master/prac6/Keras_Activation_DL_Prac_6_(ii).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [0]:
import keras
from keras import models
from keras.layers import Dense, Dropout
from keras.utils import to_categorical
from keras.datasets import mnist
from keras.utils.vis_utils import model_to_dot
from IPython.display import SVG

NUM_ROWS = 28
NUM_COLS = 28
NUM_CLASSES = 10
BATCH_SIZE = 128
EPOCHS = 10

# Load data
(X_train, y_train), (X_test, y_test) = mnist.load_data()

# Reshape data
X_train = X_train.reshape((X_train.shape[0], NUM_ROWS * NUM_COLS))
X_train = X_train.astype('float32') / 255
X_test = X_test.reshape((X_test.shape[0], NUM_ROWS * NUM_COLS))
X_test = X_test.astype('float32') / 255

# Categorically encode labels
y_train = to_categorical(y_train, NUM_CLASSES)
y_test = to_categorical(y_test, NUM_CLASSES)


# Build neural network
model = models.Sequential()
model.add(Dense(512, activation='elu', input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation='elu'))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Using TensorFlow backend.


11493376/11490434 [==============================] - 0s 0us/step





Instructions for updating:
Use tf.where in 2.0, which has the same broadcast rule as np.where



Train on 60000 samples, validate on 10000 samples
Epoch 1/10





60000/60000 [==============================] - 8s 125us/step - loss: 0.2974 - acc: 0.9093 - val_loss: 0.1448 - val_acc: 0.9557
Epoch 2/10
60000/60000 [==============================] - 7s 117us/step - loss: 0.1221 - acc: 0.9622 - val_loss: 0.1382 - val_acc: 0.9592
Epoch 3/10
60000/60000 [==============================] - 7s 120us/step - loss: 0.0834 - acc: 0.9742 - val_loss: 0.1182 - val_acc: 0.9658
Epoch 4/10
60000/60000 [==============================] - 7s 116us/step - loss: 0.0646 - acc: 0.9794 - val_loss: 0.0897 - val_acc: 0.9718
Epoch 5/10
60000/60000 [==============================] - 7s 118us/step - loss: 0.0511 - acc: 0.9835 - val_loss: 0.0876 - val_acc: 0.9745
Epoch 6/10
60000/60000 [==============================] - 7s 119us/step - loss: 0.0426 -

In [0]:
# Build neural network
model = models.Sequential()
model.add(Dense(512, activation='softmax', input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation='softmax'))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 8s 125us/step - loss: 2.2698 - acc: 0.1927 - val_loss: 2.1977 - val_acc: 0.5616
Epoch 2/10
60000/60000 [==============================] - 7s 118us/step - loss: 2.0453 - acc: 0.6449 - val_loss: 1.8609 - val_acc: 0.6600
Epoch 3/10
60000/60000 [==============================] - 7s 123us/step - loss: 1.6545 - acc: 0.6652 - val_loss: 1.4483 - val_acc: 0.6654
Epoch 4/10
60000/60000 [==============================] - 7s 122us/step - loss: 1.2689 - acc: 0.6693 - val_loss: 1.1068 - val_acc: 0.6659
Epoch 5/10
60000/60000 [==============================] - 8s 127us/step - loss: 0.9825 - acc: 0.6744 - val_loss: 0.8843 - val_acc: 0.6699
Epoch 6/10
60000/60000 [==============================] - 8s 127us/step - loss: 0.8150 - acc: 0.6832 - val_loss: 0.7731 - val_acc: 0.6838
Epoch 7/10
60000/60000 [==============================] - 8s 128us/step - loss: 0.7321 - acc: 0.6970 - val_loss: 0.7224 - 

In [0]:
# Build neural network
model = models.Sequential()
model.add(Dense(512, activation='selu', input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation='selu'))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 8s 138us/step - loss: 0.3336 - acc: 0.8996 - val_loss: 0.2138 - val_acc: 0.9302
Epoch 2/10
60000/60000 [==============================] - 8s 134us/step - loss: 0.1361 - acc: 0.9579 - val_loss: 0.1456 - val_acc: 0.9551
Epoch 3/10
60000/60000 [==============================] - 8s 133us/step - loss: 0.0976 - acc: 0.9701 - val_loss: 0.1084 - val_acc: 0.9671
Epoch 4/10
60000/60000 [==============================] - 8s 135us/step - loss: 0.0736 - acc: 0.9768 - val_loss: 0.0845 - val_acc: 0.9737
Epoch 5/10
60000/60000 [==============================] - 8s 135us/step - loss: 0.0588 - acc: 0.9815 - val_loss: 0.1073 - val_acc: 0.9697
Epoch 6/10
60000/60000 [==============================] - 8s 132us/step - loss: 0.0504 - acc: 0.9836 - val_loss: 0.0985 - val_acc: 0.9747
Epoch 7/10
60000/60000 [==============================] - 8s 133us/step - loss: 0.0419 - acc: 0.9865 - val_loss: 0.1123 - 

In [0]:
# Build neural network
model = models.Sequential()
model.add(Dense(512, activation='softplus', input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation='softplus'))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 8s 133us/step - loss: 0.4762 - acc: 0.8536 - val_loss: 0.2165 - val_acc: 0.9324
Epoch 2/10
60000/60000 [==============================] - 8s 126us/step - loss: 0.1797 - acc: 0.9449 - val_loss: 0.1226 - val_acc: 0.9635
Epoch 3/10
60000/60000 [==============================] - 8s 125us/step - loss: 0.1170 - acc: 0.9646 - val_loss: 0.1112 - val_acc: 0.9666
Epoch 4/10
60000/60000 [==============================] - 8s 126us/step - loss: 0.0859 - acc: 0.9731 - val_loss: 0.0909 - val_acc: 0.9731
Epoch 5/10
60000/60000 [==============================] - 7s 124us/step - loss: 0.0674 - acc: 0.9792 - val_loss: 0.0975 - val_acc: 0.9715
Epoch 6/10
60000/60000 [==============================] - 8s 126us/step - loss: 0.0549 - acc: 0.9829 - val_loss: 0.0714 - val_acc: 0.9789
Epoch 7/10
60000/60000 [==============================] - 8s 128us/step - loss: 0.0441 - acc: 0.9860 - val_loss: 0.0756 - 

In [0]:
# Build neural network
model = models.Sequential()
model.add(Dense(512, activation='softsign', input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation='softsign'))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 7s 124us/step - loss: 0.2980 - acc: 0.9110 - val_loss: 0.1795 - val_acc: 0.9464
Epoch 2/10
60000/60000 [==============================] - 7s 118us/step - loss: 0.1478 - acc: 0.9557 - val_loss: 0.1263 - val_acc: 0.9609
Epoch 3/10
60000/60000 [==============================] - 7s 115us/step - loss: 0.1023 - acc: 0.9691 - val_loss: 0.1054 - val_acc: 0.9670
Epoch 4/10
60000/60000 [==============================] - 7s 118us/step - loss: 0.0764 - acc: 0.9769 - val_loss: 0.0885 - val_acc: 0.9723
Epoch 5/10
60000/60000 [==============================] - 7s 120us/step - loss: 0.0582 - acc: 0.9825 - val_loss: 0.0802 - val_acc: 0.9741
Epoch 6/10
60000/60000 [==============================] - 7s 119us/step - loss: 0.0450 - acc: 0.9862 - val_loss: 0.0680 - val_acc: 0.9778
Epoch 7/10
60000/60000 [==============================] - 7s 121us/step - loss: 0.0347 - acc: 0.9890 - val_loss: 0.0747 - 

In [0]:
# Build neural network
model = models.Sequential()
model.add(Dense(512, activation='relu', input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation='relu'))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 7s 122us/step - loss: 0.2329 - acc: 0.9290 - val_loss: 0.1027 - val_acc: 0.9665
Epoch 2/10
60000/60000 [==============================] - 7s 118us/step - loss: 0.0867 - acc: 0.9733 - val_loss: 0.0799 - val_acc: 0.9754
Epoch 3/10
60000/60000 [==============================] - 7s 117us/step - loss: 0.0557 - acc: 0.9827 - val_loss: 0.0863 - val_acc: 0.9758
Epoch 4/10
60000/60000 [==============================] - 7s 118us/step - loss: 0.0407 - acc: 0.9868 - val_loss: 0.0794 - val_acc: 0.9771
Epoch 5/10
60000/60000 [==============================] - 7s 119us/step - loss: 0.0312 - acc: 0.9899 - val_loss: 0.0767 - val_acc: 0.9801
Epoch 6/10
60000/60000 [==============================] - 7s 118us/step - loss: 0.0231 - acc: 0.9926 - val_loss: 0.0812 - val_acc: 0.9799
Epoch 7/10
60000/60000 [==============================] - 7s 121us/step - loss: 0.0186 - acc: 0.9940 - val_loss: 0.0844 - 

In [0]:
# Build neural network
model = models.Sequential()
model.add(Dense(512, activation='tanh', input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation='tanh'))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 8s 126us/step - loss: 0.2874 - acc: 0.9132 - val_loss: 0.1749 - val_acc: 0.9477
Epoch 2/10
60000/60000 [==============================] - 7s 118us/step - loss: 0.1224 - acc: 0.9629 - val_loss: 0.1091 - val_acc: 0.9676
Epoch 3/10
60000/60000 [==============================] - 7s 118us/step - loss: 0.0823 - acc: 0.9746 - val_loss: 0.0913 - val_acc: 0.9706
Epoch 4/10
60000/60000 [==============================] - 7s 119us/step - loss: 0.0592 - acc: 0.9822 - val_loss: 0.0814 - val_acc: 0.9758
Epoch 5/10
60000/60000 [==============================] - 7s 119us/step - loss: 0.0431 - acc: 0.9866 - val_loss: 0.0812 - val_acc: 0.9748
Epoch 6/10
60000/60000 [==============================] - 7s 116us/step - loss: 0.0319 - acc: 0.9905 - val_loss: 0.0656 - val_acc: 0.9796
Epoch 7/10
60000/60000 [==============================] - 7s 119us/step - loss: 0.0230 - acc: 0.9936 - val_loss: 0.0638 - 

In [0]:
# Build neural network
model = models.Sequential()
model.add(Dense(512, activation='sigmoid', input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation='sigmoid'))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 8s 129us/step - loss: 0.4875 - acc: 0.8598 - val_loss: 0.2726 - val_acc: 0.9147
Epoch 2/10
60000/60000 [==============================] - 7s 119us/step - loss: 0.2170 - acc: 0.9352 - val_loss: 0.1879 - val_acc: 0.9435
Epoch 3/10
60000/60000 [==============================] - 7s 120us/step - loss: 0.1561 - acc: 0.9533 - val_loss: 0.1371 - val_acc: 0.9585
Epoch 4/10
60000/60000 [==============================] - 7s 117us/step - loss: 0.1180 - acc: 0.9640 - val_loss: 0.1099 - val_acc: 0.9654
Epoch 5/10
60000/60000 [==============================] - 7s 118us/step - loss: 0.0921 - acc: 0.9720 - val_loss: 0.1011 - val_acc: 0.9693
Epoch 6/10
60000/60000 [==============================] - 7s 119us/step - loss: 0.0744 - acc: 0.9769 - val_loss: 0.0946 - val_acc: 0.9718
Epoch 7/10
60000/60000 [==============================] - 7s 119us/step - loss: 0.0630 - acc: 0.9805 - val_loss: 0.0934 - 

In [0]:
# Build neural network
model = models.Sequential()
model.add(Dense(512, activation='hard_sigmoid', input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation='hard_sigmoid'))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 8s 137us/step - loss: 0.4983 - acc: 0.8548 - val_loss: 0.2510 - val_acc: 0.9225
Epoch 2/10
60000/60000 [==============================] - 7s 124us/step - loss: 0.2138 - acc: 0.9355 - val_loss: 0.1724 - val_acc: 0.9481
Epoch 3/10
60000/60000 [==============================] - 8s 126us/step - loss: 0.1494 - acc: 0.9549 - val_loss: 0.1327 - val_acc: 0.9594
Epoch 4/10
60000/60000 [==============================] - 7s 123us/step - loss: 0.1117 - acc: 0.9663 - val_loss: 0.1140 - val_acc: 0.9648
Epoch 5/10
60000/60000 [==============================] - 8s 126us/step - loss: 0.0868 - acc: 0.9736 - val_loss: 0.0954 - val_acc: 0.9702
Epoch 6/10
60000/60000 [==============================] - 8s 127us/step - loss: 0.0693 - acc: 0.9792 - val_loss: 0.0876 - val_acc: 0.9730
Epoch 7/10
60000/60000 [==============================] - 8s 125us/step - loss: 0.0573 - acc: 0.9824 - val_loss: 0.0916 - 

In [0]:
# Build neural network
model = models.Sequential()
model.add(Dense(512, activation='exponential', input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation='exponential'))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 8s 130us/step - loss: 14.5378 - acc: 0.0973 - val_loss: 14.5353 - val_acc: 0.0982
Epoch 2/10
60000/60000 [==============================] - 7s 116us/step - loss: 14.5487 - acc: 0.0974 - val_loss: 14.5353 - val_acc: 0.0982
Epoch 3/10
60000/60000 [==============================] - 7s 117us/step - loss: 14.5487 - acc: 0.0974 - val_loss: 14.5353 - val_acc: 0.0982
Epoch 4/10
60000/60000 [==============================] - 7s 119us/step - loss: 14.5487 - acc: 0.0974 - val_loss: 14.5353 - val_acc: 0.0982
Epoch 5/10
60000/60000 [==============================] - 7s 119us/step - loss: 14.5487 - acc: 0.0974 - val_loss: 14.5353 - val_acc: 0.0982
Epoch 6/10
60000/60000 [==============================] - 7s 119us/step - loss: 14.5487 - acc: 0.0974 - val_loss: 14.5353 - val_acc: 0.0982
Epoch 7/10
60000/60000 [==============================] - 7s 121us/step - loss: 14.5487 - acc: 0.0974 - val_lo

In [0]:
# Build neural network
model = models.Sequential()
model.add(Dense(512, activation='linear', input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation='linear'))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 8s 136us/step - loss: 0.4310 - acc: 0.8752 - val_loss: 0.3593 - val_acc: 0.8976
Epoch 2/10
60000/60000 [==============================] - 7s 119us/step - loss: 0.3463 - acc: 0.9026 - val_loss: 0.3316 - val_acc: 0.9042
Epoch 3/10
60000/60000 [==============================] - 7s 118us/step - loss: 0.3235 - acc: 0.9096 - val_loss: 0.3022 - val_acc: 0.9178
Epoch 4/10
60000/60000 [==============================] - 7s 119us/step - loss: 0.3124 - acc: 0.9122 - val_loss: 0.3459 - val_acc: 0.9060
Epoch 5/10
60000/60000 [==============================] - 7s 118us/step - loss: 0.3023 - acc: 0.9159 - val_loss: 0.3242 - val_acc: 0.9044
Epoch 6/10
60000/60000 [==============================] - 7s 116us/step - loss: 0.2934 - acc: 0.9185 - val_loss: 0.2979 - val_acc: 0.9186
Epoch 7/10
60000/60000 [==============================] - 7s 118us/step - loss: 0.2917 - acc: 0.9187 - val_loss: 0.3268 - 

In [0]:
# Build neural network
model = models.Sequential()
model.add(Dense(512, activation=keras.layers.LeakyReLU(alpha=0.3), input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation=keras.layers.LeakyReLU(alpha=0.3)))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

/usr/local/lib/python3.6/dist-packages/keras/activations.py:235: UserWarning: Do not pass a layer instance (such as LeakyReLU) as the activation argument of another layer. Instead, advanced activation layers should be used just like any other layer in a model.
  identifier=identifier.__class__.__name__))


Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 8s 136us/step - loss: 0.2683 - acc: 0.9177 - val_loss: 0.1620 - val_acc: 0.9487
Epoch 2/10
60000/60000 [==============================] - 7s 122us/step - loss: 0.1274 - acc: 0.9613 - val_loss: 0.1178 - val_acc: 0.9644
Epoch 3/10
60000/60000 [==============================] - 7s 121us/step - loss: 0.0930 - acc: 0.9710 - val_loss: 0.1070 - val_acc: 0.9705
Epoch 4/10
60000/60000 [==============================] - 7s 121us/step - loss: 0.0743 - acc: 0.9764 - val_loss: 0.1195 - val_acc: 0.9676
Epoch 5/10
60000/60000 [==============================] - 7s 121us/step - loss: 0.0638 - acc: 0.9799 - val_loss: 0.0821 - val_acc: 0.9760
Epoch 6/10
60000/60000 [==============================] - 7s 121us/step - loss: 0.0562 - acc: 0.9829 - val_loss: 0.0947 - val_acc: 0.9735
Epoch 7/10
60000/60000 [==============================] - 7s 119us/step - loss: 0.0483 - acc: 0.9844 - val_loss: 0.0830 - 

In [0]:
# Build neural network
model = models.Sequential()
model.add(Dense(512, activation=keras.layers.PReLU(alpha_initializer='zeros', alpha_regularizer=None, alpha_constraint=None, shared_axes=None)
, input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation=keras.layers.PReLU(alpha_initializer='zeros', alpha_regularizer=None, alpha_constraint=None, shared_axes=None)
))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

/usr/local/lib/python3.6/dist-packages/keras/activations.py:235: UserWarning: Do not pass a layer instance (such as PReLU) as the activation argument of another layer. Instead, advanced activation layers should be used just like any other layer in a model.
  identifier=identifier.__class__.__name__))


Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 9s 142us/step - loss: 0.2251 - acc: 0.9310 - val_loss: 0.1383 - val_acc: 0.9549
Epoch 2/10
60000/60000 [==============================] - 8s 126us/step - loss: 0.0850 - acc: 0.9745 - val_loss: 0.0891 - val_acc: 0.9718
Epoch 3/10
60000/60000 [==============================] - 8s 125us/step - loss: 0.0568 - acc: 0.9823 - val_loss: 0.0829 - val_acc: 0.9756
Epoch 4/10
60000/60000 [==============================] - 8s 126us/step - loss: 0.0410 - acc: 0.9871 - val_loss: 0.0768 - val_acc: 0.9796
Epoch 5/10
60000/60000 [==============================] - 8s 126us/step - loss: 0.0311 - acc: 0.9903 - val_loss: 0.0873 - val_acc: 0.9763
Epoch 6/10
60000/60000 [==============================] - 8s 126us/step - loss: 0.0245 - acc: 0.9923 - val_loss: 0.0793 - val_acc: 0.9820
Epoch 7/10
60000/60000 [==============================] - 8s 126us/step - loss: 0.0190 - acc: 0.9939 - val_loss: 0.0811 - 

In [0]:
# Build neural network
model = models.Sequential()
model.add(Dense(512, activation=keras.layers.ELU(alpha=1.0), input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation=keras.layers.ELU(alpha=1.0)))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])


/usr/local/lib/python3.6/dist-packages/keras/activations.py:235: UserWarning: Do not pass a layer instance (such as ELU) as the activation argument of another layer. Instead, advanced activation layers should be used just like any other layer in a model.
  identifier=identifier.__class__.__name__))


Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 8s 138us/step - loss: 0.2874 - acc: 0.9126 - val_loss: 0.2173 - val_acc: 0.9319
Epoch 2/10
60000/60000 [==============================] - 7s 119us/step - loss: 0.1243 - acc: 0.9618 - val_loss: 0.1177 - val_acc: 0.9599
Epoch 3/10
60000/60000 [==============================] - 7s 122us/step - loss: 0.0852 - acc: 0.9733 - val_loss: 0.0938 - val_acc: 0.9714
Epoch 4/10
60000/60000 [==============================] - 7s 123us/step - loss: 0.0646 - acc: 0.9794 - val_loss: 0.1050 - val_acc: 0.9703
Epoch 5/10
60000/60000 [==============================] - 7s 123us/step - loss: 0.0517 - acc: 0.9836 - val_loss: 0.0912 - val_acc: 0.9716
Epoch 6/10
60000/60000 [==============================] - 7s 123us/step - loss: 0.0443 - acc: 0.9856 - val_loss: 0.0797 - val_acc: 0.9779
Epoch 7/10
60000/60000 [==============================] - 7s 122us/step - loss: 0.0353 - acc: 0.9891 - val_loss: 0.0869 - 

In [0]:
# Build neural network
model = models.Sequential()
model.add(Dense(512, activation=keras.layers.ThresholdedReLU(theta=1.0), input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation=keras.layers.ThresholdedReLU(theta=1.0)))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

/usr/local/lib/python3.6/dist-packages/keras/activations.py:235: UserWarning: Do not pass a layer instance (such as ThresholdedReLU) as the activation argument of another layer. Instead, advanced activation layers should be used just like any other layer in a model.
  identifier=identifier.__class__.__name__))


Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 9s 146us/step - loss: 0.4720 - acc: 0.8510 - val_loss: 0.1701 - val_acc: 0.9477
Epoch 2/10
60000/60000 [==============================] - 8s 135us/step - loss: 0.1466 - acc: 0.9561 - val_loss: 0.1284 - val_acc: 0.9613
Epoch 3/10
60000/60000 [==============================] - 7s 123us/step - loss: 0.0990 - acc: 0.9703 - val_loss: 0.0897 - val_acc: 0.9714
Epoch 4/10
60000/60000 [==============================] - 7s 124us/step - loss: 0.0755 - acc: 0.9767 - val_loss: 0.0913 - val_acc: 0.9720
Epoch 5/10
60000/60000 [==============================] - 7s 125us/step - loss: 0.0608 - acc: 0.9812 - val_loss: 0.0842 - val_acc: 0.9758
Epoch 6/10
60000/60000 [==============================] - 7s 123us/step - loss: 0.0500 - acc: 0.9842 - val_loss: 0.0743 - val_acc: 0.9770
Epoch 7/10
60000/60000 [==============================] - 7s 123us/step - loss: 0.0412 - acc: 0.9872 - val_loss: 0.0774 - 

In [0]:
# Build neural network
model = models.Sequential()
model.add(Dense(512, activation=keras.layers.Softmax(axis=-1), input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation=keras.layers.Softmax(axis=-1)))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

/usr/local/lib/python3.6/dist-packages/keras/activations.py:235: UserWarning: Do not pass a layer instance (such as Softmax) as the activation argument of another layer. Instead, advanced activation layers should be used just like any other layer in a model.
  identifier=identifier.__class__.__name__))


Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 9s 143us/step - loss: 2.2697 - acc: 0.1997 - val_loss: 2.1976 - val_acc: 0.6927
Epoch 2/10
60000/60000 [==============================] - 7s 123us/step - loss: 2.0456 - acc: 0.7670 - val_loss: 1.8603 - val_acc: 0.8196
Epoch 3/10
60000/60000 [==============================] - 7s 122us/step - loss: 1.6499 - acc: 0.8201 - val_loss: 1.4316 - val_acc: 0.8294
Epoch 4/10
60000/60000 [==============================] - 7s 122us/step - loss: 1.2309 - acc: 0.8293 - val_loss: 1.0370 - val_acc: 0.8337
Epoch 5/10
60000/60000 [==============================] - 7s 123us/step - loss: 0.8835 - acc: 0.8335 - val_loss: 0.7435 - val_acc: 0.8381
Epoch 6/10
60000/60000 [==============================] - 7s 119us/step - loss: 0.6490 - acc: 0.8401 - val_loss: 0.5694 - val_acc: 0.8492
Epoch 7/10
60000/60000 [==============================] - 7s 121us/step - loss: 0.5219 - acc: 0.8497 - val_loss: 0.4822 - 

In [0]:
# Build neural network
model = models.Sequential()
model.add(Dense(512, activation=keras.layers.ReLU(max_value=None, negative_slope=0.0, threshold=0.0)
, input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation=keras.layers.ReLU(max_value=None, negative_slope=0.0, threshold=0.0)
))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

/usr/local/lib/python3.6/dist-packages/keras/activations.py:235: UserWarning: Do not pass a layer instance (such as ReLU) as the activation argument of another layer. Instead, advanced activation layers should be used just like any other layer in a model.
  identifier=identifier.__class__.__name__))


Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 8s 137us/step - loss: 0.2272 - acc: 0.9303 - val_loss: 0.1032 - val_acc: 0.9683
Epoch 2/10
60000/60000 [==============================] - 7s 117us/step - loss: 0.0858 - acc: 0.9737 - val_loss: 0.0846 - val_acc: 0.9739
Epoch 3/10
60000/60000 [==============================] - 7s 118us/step - loss: 0.0565 - acc: 0.9828 - val_loss: 0.0747 - val_acc: 0.9778
Epoch 4/10
60000/60000 [==============================] - 7s 116us/step - loss: 0.0403 - acc: 0.9870 - val_loss: 0.0678 - val_acc: 0.9803
Epoch 5/10
60000/60000 [==============================] - 7s 118us/step - loss: 0.0297 - acc: 0.9904 - val_loss: 0.0828 - val_acc: 0.9783
Epoch 6/10
60000/60000 [==============================] - 7s 120us/step - loss: 0.0236 - acc: 0.9927 - val_loss: 0.0743 - val_acc: 0.9808
Epoch 7/10
60000/60000 [==============================] - 7s 120us/step - loss: 0.0185 - acc: 0.9943 - val_loss: 0.0850 - 

relu

Test loss: 0.08459654634964409

Test accuracy: 0.9799

softmax

Test loss: 0.39845509810447693

Test accuracy: 0.8642

selu

Test loss: 0.11848029121595427

Test accuracy: 0.9726

softplus

Test loss: 0.08013557475531798

Test accuracy: 0.9798

softsign

Test loss: 0.07156896027913608

Test accuracy: 0.9793

relu

Test loss: 0.09273777272929686

Test accuracy: 0.9843

tanh

Test loss: 0.06885148700817954

Test accuracy: 0.9803

sigmoid

Test loss: 0.07136517209208104

Test accuracy: 0.9781

hard_sigmoid

Test loss: 0.06968761095895898

Test accuracy: 0.9795

exponential

Test loss: 14.573981648254394

Test accuracy: 0.0958

linear

Test loss: 0.30165807257443666

Test accuracy: 0.9158

LeakyReLU

Test loss: 0.11553157684192265

Test accuracy: 0.9709

PReLU

Test loss: 0.11553157684192265

Test accuracy: 0.9709

ELU

Test loss: 0.1115254690002792

Test accuracy: 0.9743

ThresholdedReLU

Test loss: 0.08483904730657359

Test accuracy: 0.9786

Softmax

Test loss: 0.2818538703083992

Test accuracy: 0.9333

ReLU

Test loss: 0.09205757192509677

Test accuracy: 0.9829

In [0]:
import keras
from keras import models
from keras.layers import Dense, Dropout
from keras.utils import to_categorical
from keras.datasets import fashion_mnist
from keras.utils.vis_utils import model_to_dot
from IPython.display import SVG
from keras.layers import Activation, Flatten, Conv2D

NUM_ROWS = 28
NUM_COLS = 28
NUM_CLASSES = 10
BATCH_SIZE = 128
EPOCHS = 10

# Load data
(X_train, y_train), (X_test, y_test) = fashion_mnist.load_data()

# Reshape data
X_train = X_train.reshape((X_train.shape[0], 28, 28, 1))
X_train = X_train.astype('float32') / 255
X_test = X_test.reshape((X_test.shape[0], 28, 28, 1))
X_test = X_test.astype('float32') / 255



# Categorically encode labels
y_train = to_categorical(y_train, NUM_CLASSES)
y_test = to_categorical(y_test, NUM_CLASSES)

# Build neural network
model = models.Sequential()
model.add(Conv2D(512, (28, 28), padding="valid", activation=keras.layers.ReLU(max_value=None, negative_slope=0.0, threshold=0.0), input_shape=X_train.shape[1:]))
model.add(Flatten())
model.add(Dense(512, activation=keras.layers.ReLU(max_value=None, negative_slope=0.0, threshold=0.0)
, input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation=keras.layers.ReLU(max_value=None, negative_slope=0.0, threshold=0.0)
))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

4423680/4422102 [==============================] - 1s 0us/step


/usr/local/lib/python3.6/dist-packages/keras/activations.py:235: UserWarning: Do not pass a layer instance (such as ReLU) as the activation argument of another layer. Instead, advanced activation layers should be used just like any other layer in a model.
  identifier=identifier.__class__.__name__))


Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 14s 236us/step - loss: 0.6057 - acc: 0.7746 - val_loss: 0.5649 - val_acc: 0.7894
Epoch 2/10
60000/60000 [==============================] - 13s 219us/step - loss: 0.4012 - acc: 0.8529 - val_loss: 0.3977 - val_acc: 0.8574
Epoch 3/10
60000/60000 [==============================] - 13s 217us/step - loss: 0.3561 - acc: 0.8698 - val_loss: 0.3550 - val_acc: 0.8758
Epoch 4/10
60000/60000 [==============================] - 13s 225us/step - loss: 0.3288 - acc: 0.8794 - val_loss: 0.3850 - val_acc: 0.8621
Epoch 5/10
60000/60000 [==============================] - 13s 221us/step - loss: 0.3146 - acc: 0.8837 - val_loss: 0.4231 - val_acc: 0.8606
Epoch 6/10
60000/60000 [==============================] - 13s 221us/step - loss: 0.2994 - acc: 0.8905 - val_loss: 0.3554 - val_acc: 0.8788
Epoch 7/10
60000/60000 [==============================] - 15s 244us/step - loss: 0.2921 - acc: 0.8923 - val_loss: 0.

FASHION_MNIST

Test loss: 0.4138208844304085

Test accuracy: 0.8699

In [0]:
import keras
from keras import models
from keras.layers import Dense, Dropout, Conv2D,Flatten
from keras.utils import to_categorical
from keras.datasets import cifar10
from keras.utils.vis_utils import model_to_dot
from IPython.display import SVG
from keras.preprocessing.image import ImageDataGenerator
from keras import optimizers

#import dataset
(X_train, y_train), (X_test, y_test) = cifar10.load_data()

#change shape from image to vector
X_train = X_train.reshape(50000,32,32,3)
X_test = X_test.reshape(10000,32,32,3)

#preprocess
X_train = X_train.astype('float32')
X_test = X_test.astype('float32')
X_train /= 255.0
X_test /= 255.0

#change labels from numeric to one hot encoded
Y_train = to_categorical(y_train, 10)
Y_test =  to_categorical(y_test, 10)

adamax=keras.optimizers.Adamax(lr=0.002, beta_1=0.9, beta_2=0.999)


# Build neural network
model = models.Sequential()
model.add(Conv2D(1024, (32, 32), padding="valid", activation=keras.layers.ReLU(max_value=None, negative_slope=0.0, threshold=0.0), input_shape=X_train.shape[1:]))
model.add(Flatten())
model.add(Dense(512, activation=keras.layers.ReLU(max_value=None, negative_slope=0.0, threshold=0.0)
, input_shape=(32 * 32,)))
model.add(Dense(256, activation=keras.layers.ReLU(max_value=None, negative_slope=0.0, threshold=0.0)
))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='adamax',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, Y_train,
          batch_size=128,
          epochs=10,
          verbose=1,
          validation_data=(X_test, Y_test))

score = model.evaluate(X_test, Y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Using TensorFlow backend.


170500096/170498071 [==============================] - 11s 0us/step





Instructions for updating:
Use tf.where in 2.0, which has the same broadcast rule as np.where


/usr/local/lib/python3.6/dist-packages/keras/activations.py:235: UserWarning: Do not pass a layer instance (such as ReLU) as the activation argument of another layer. Instead, advanced activation layers should be used just like any other layer in a model.
  identifier=identifier.__class__.__name__))





Train on 50000 samples, validate on 10000 samples
Epoch 1/10





50000/50000 [==============================] - 17s 342us/step - loss: 2.0190 - acc: 0.2931 - val_loss: 1.7905 - val_acc: 0.3582
Epoch 2/10
50000/50000 [==============================] - 3s 60us/step - loss: 1.7160 - acc: 0.3868 - val_loss: 1.6572 - val_acc: 0.3980
Epoch 3/10
50000/50000 [==============================] - 3s 59us/step - loss: 1.6237 - acc: 0.4200 - val_loss: 1.6376 - val_acc: 0.4148
Epoch 4/10
50000/50000 [==============================] - 3s 60us/step - loss: 1.5452 - acc: 0.4490 - val_loss: 1.5281 - val_acc: 0.4622
Epoch 5/10
50000/50000 [==============================] - 3s 60us/step - loss: 1.4916 - acc: 0.4685 - val_loss: 1.5196 - val_acc: 0.4611
Epoch 6/10
50000/50000 [==============================] - 3s 61us/step - loss: 1.4501 - acc: 0.4804 - val_loss: 1.4704 - val_acc: 0.4771
Epoch 7/10
50000/50000 [==============================] - 3s 61us/step - loss: 1.4051 - acc: 0.4999 - val_loss: 1.4850

CIFAR10

Test loss: 1.4009930366516112

Test accuracy: 0.4971

In [0]:
import keras
from keras import models
from keras.layers import Dense, Dropout, Conv2D,Flatten
from keras.utils import to_categorical
from keras.datasets import cifar100
from keras.utils.vis_utils import model_to_dot
from IPython.display import SVG
from keras.preprocessing.image import ImageDataGenerator
from keras import optimizers

#import dataset
(X_train, y_train), (X_test, y_test) = cifar100.load_data()

#change shape from image to vector
X_train = X_train.reshape(50000,32,32,3)
X_test = X_test.reshape(10000,32,32,3)

#preprocess
X_train = X_train.astype('float32')
X_test = X_test.astype('float32')
X_train /= 255.0
X_test /= 255.0

#change labels from numeric to one hot encoded
Y_train = to_categorical(y_train, 100)
Y_test =  to_categorical(y_test, 100)

adamax=keras.optimizers.Adamax(lr=0.002, beta_1=0.9, beta_2=0.999)


# Build neural network
model = models.Sequential()
model.add(Conv2D(1024, (32, 32), padding="valid", activation=keras.layers.ReLU(max_value=None, negative_slope=0.0, threshold=0.0), input_shape=X_train.shape[1:]))
model.add(Flatten())
model.add(Dense(512, activation=keras.layers.ReLU(max_value=None, negative_slope=0.0, threshold=0.0)
, input_shape=(32 * 32,)))
model.add(Dense(256, activation=keras.layers.ReLU(max_value=None, negative_slope=0.0, threshold=0.0)
))
model.add(Dense(100, activation='softmax'))

# Compile model
model.compile(optimizer='adamax',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, Y_train,
          batch_size=128,
          epochs=10,
          verbose=1,
          validation_data=(X_test, Y_test))

score = model.evaluate(X_test, Y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

169009152/169001437 [==============================] - 11s 0us/step


/usr/local/lib/python3.6/dist-packages/keras/activations.py:235: UserWarning: Do not pass a layer instance (such as ReLU) as the activation argument of another layer. Instead, advanced activation layers should be used just like any other layer in a model.
  identifier=identifier.__class__.__name__))


Train on 50000 samples, validate on 10000 samples
Epoch 1/10
50000/50000 [==============================] - 3s 69us/step - loss: 4.3009 - acc: 0.0522 - val_loss: 3.9769 - val_acc: 0.0947
Epoch 2/10
50000/50000 [==============================] - 3s 63us/step - loss: 3.8355 - acc: 0.1129 - val_loss: 3.7477 - val_acc: 0.1272
Epoch 3/10
50000/50000 [==============================] - 3s 63us/step - loss: 3.6404 - acc: 0.1462 - val_loss: 3.6055 - val_acc: 0.1534
Epoch 4/10
50000/50000 [==============================] - 3s 61us/step - loss: 3.5076 - acc: 0.1701 - val_loss: 3.4881 - val_acc: 0.1786
Epoch 5/10
50000/50000 [==============================] - 3s 62us/step - loss: 3.3982 - acc: 0.1887 - val_loss: 3.4474 - val_acc: 0.1816
Epoch 6/10
50000/50000 [==============================] - 3s 63us/step - loss: 3.3026 - acc: 0.2066 - val_loss: 3.3710 - val_acc: 0.2018
Epoch 7/10
50000/50000 [==============================] - 3s 62us/step - loss: 3.2143 - acc: 0.2225 - val_loss: 3.3112 - val_acc

CIFAR100

Test loss: 3.242277897644043

Test accuracy: 0.2326